base-theme/assets/webpack/webpack.common.ts

```diff
@@ -141,12 +141,12 @@ const config: webpack.Configuration = {
     new CopyWebpackPlugin({
       patterns: [
         {
-          from:        "./base-theme/assets/fonts/**/*.{ttf,woff,woff2}",
+          from:        "./base-theme/assets/static_shared/fonts/*.{ttf,woff,woff2}",
```

Comment:
After running the theme assets pipeline against this branch and subsequently building a site against it, there are several broken assets:

![image](https://github.com/mitodl/ocw-hugo-themes/assets/12089658/bff3535b-f389-4ebe-a019-40fd1d059660)

Starting with fonts, the webpack manifest seems to have the hash written into the filename, which is why the page is attempting to load the files that way. However, looking at the output of the theme assets build against this branch, I see the following in `upload-theme-assets-task`:

```
...
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/fonts/22d194ee0b49f6e9d633ddc302f108b1.woff2 to s3://ocw-content-draft/static_shared/css/fonts/22d194ee0b49f6e9d633ddc302f108b1.woff2
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/course_offline.150d9.css to s3://ocw-content-draft/static_shared/css/course_offline.150d9.css
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/fonts/bbe2fbc76615c7cbd85ea132d0ff92e7.woff to s3://ocw-content-draft/static_shared/css/fonts/bbe2fbc76615c7cbd85ea132d0ff92e7.woff
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/course_v2.150d9.css to s3://ocw-content-draft/static_shared/css/course_v2.150d9.css
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/fields.150d9.css to s3://ocw-content-draft/static_shared/css/fields.150d9.css
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/MaterialIcons-Regular.subset.woff to s3://ocw-content-draft/static_shared/fonts/MaterialIcons-Regular.subset.woff
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/MaterialIcons-Regular.subset.ttf to s3://ocw-content-draft/static_shared/fonts/MaterialIcons-Regular.subset.ttf
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/Helvetica-Light.ttf to s3://ocw-content-draft/static_shared/fonts/Helvetica-Light.ttf
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/Cardo-Italic.ttf to s3://ocw-content-draft/static_shared/fonts/Cardo-Italic.ttf
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/MaterialIcons-Regular.subset.woff2 to s3://ocw-content-draft/static_shared/fonts/MaterialIcons-Regular.subset.woff2
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/Cardo-Bold.ttf to s3://ocw-content-draft/static_shared/fonts/Cardo-Bold.ttf
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/www.150d9.css to s3://ocw-content-draft/static_shared/css/www.150d9.css
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/css/www_offline.150d9.css to s3://ocw-content-draft/static_shared/css/www_offline.150d9.css
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/Cardo-Regular.ttf to s3://ocw-content-draft/static_shared/fonts/Cardo-Regular.ttf
upload: ocw-hugo-themes-git/base-theme/dist/static_shared/fonts/MaterialIconsRound-Regular.subset.woff2 to s3://ocw-content-draft/static_shared/fonts/MaterialIconsRound-Regular.subset.woff2
...
```

As you can see, Webpack seems to be generating both an unhashed filename as well as a filename with just the hash. The hashes in the webpack manifest being requested by the rendered site also don't seem to match any of the "hash only" filenames. For example, if I take `MaterialIcons-Regular.subset.08116c5bd8d6472837e88e4940e829b7.woff` and take `08116c5bd8d6472837e88e4940e829b7` out of that and search the theme assets build output, I don't see it anywhere. The same seems to be happening for some images.

The theme assets pipeline simply runs `npm run build:webpack`, and then uploads the contents of the `dist` and `static` folders to both the draft and live buckets. The `webpack.json` file is uploaded to `ol-eng-artifacts` under `ocw-hugo-themes/branch/webpack.json`, where `branch` is whatever branch of `ocw-hugo-themes` was built. The site pipeline pulls down the `webpack.json` file and injects it into the build, using the paths contained within to generate URLs during the Hugo build. The `webpack.json` that was built by this branch looks like this:

```json
{
  "course_v2": {
    "css": "/static_shared/css/course_v2.150d9.css",
    "js": "/static_shared/js/course_v2.150d9.js"
  },
  "course_offline": {
    "css": "/static_shared/css/course_offline.150d9.css",
    "js": "/static_shared/js/course_offline.150d9.js"
  },
  "instructor_insights": {
    "css": "/static_shared/css/instructor_insights.150d9.css",
    "js": "/static_shared/js/instructor_insights.150d9.js"
  },
  "www": {
    "css": "/static_shared/css/www.150d9.css",
    "js": "/static_shared/js/www.150d9.js"
  },
  "www_offline": {
    "css": "/static_shared/css/www_offline.150d9.css",
    "js": "/static_shared/js/www_offline.150d9.js"
  },
  "fields": {
    "css": "/static_shared/css/fields.150d9.css",
    "js": "/static_shared/js/fields.150d9.js"
  },
  "common": {
    "js": "/static_shared/js/common.150d9.js"
  },
  "": {
    "js": [
      "/static_shared/mathjax/tex-svg-full.d4bed8ec220a6d2f41f9.js",
      "/static_shared/mathjax/tex-mml-svg.4eb9c62f95eb8f4db3f5.js",
      "/static_shared/mathjax/tex-svg.1e5aac33551eab003199.js",
      "/static_shared/mathjax/tex-chtml-full-speech.6b4339b64cc61ce8e22b.js",
      "/static_shared/mathjax/mml-svg.8e8139f345d391e913d6.js",
      "/static_shared/mathjax/tex-chtml-full.8163349bfe97c57b2798.js",
      "/static_shared/mathjax/tex-mml-chtml.22550b680214c18eff35.js",
      "/static_shared/mathjax/tex-chtml.f6ecc8299ee514404dda.js",
      "/static_shared/mathjax/tex.4bf0b1adde4b282f6e9f.js",
      "/static_shared/mathjax/mml-chtml.273ff33276facf56f2a2.js",
      "/static_shared/js/238.150d9.js",
      "/static_shared/mathjax/menu.b3dd0d8362ebef9c63af.js",
      "/static_shared/mathjax/tex-full.0c94c110dd67c8943575.js",
      "/static_shared/mathjax/sre.56a06167a3efb36de444.js",
      "/static_shared/mathjax/node-main.54b9dfcb3a3ca3ccdf31.js",
      "/static_shared/mathjax/chtml.def31f8fefae1a84af39.js",
      "/static_shared/mathjax/core.59c8aab98ac84e64698d.js",
      "/static_shared/mathjax/all-packages.c43f30ebcf5d055d52c6.js",
      "/static_shared/mathjax/svg.a162e7ee4f0b75e4e1e1.js",
      "/static_shared/js/948.150d9.js",
      "/static_shared/mathjax/tex.774eb9affe6653bbe816.js",
      "/static_shared/mathjax/tex-base.79d1f48ba8ed2a27a445.js",
      "/static_shared/mathjax/asciimath.0ff3f7edd104e7fb8126.js",
      "/static_shared/mathjax/tex.d995272e24916f40d3d9.js",
      "/static_shared/mathjax/entities.6ac9844259594c3d664e.js",
      "/static_shared/mathjax/mhchem.5a1ab5602e7dc5fad117.js",
      "/static_shared/mathjax/explorer.3899f1be0d362d27694f.js",
      "/static_shared/mathjax/mathtools.829a113c19fd1c92207e.js",
      "/static_shared/mathjax/startup.7690c60c343c5d8e58dc.js",
      "/static_shared/mathjax/mml3.87049420b71d06f11952.js",
      "/static_shared/mathjax/ams.87856963d03f3178232d.js",
      "/static_shared/mathjax/physics.17f7cd1d0c856d616864.js",
      "/static_shared/mathjax/liteDOM.852365260d533d5bfa7d.js",
      "/static_shared/mathjax/loader.2f7e9c96d0396612ecbb.js",
      "/static_shared/mathjax/bussproofs.aec0fe070aa2d9266e26.js",
      "/static_shared/mathjax/complexity.b6752b7f616d7b68d15c.js",
      "/static_shared/mathjax/textmacros.5aa515b6a28cb523b195.js",
      "/static_shared/mathjax/mml.0082d9d08da1feaa3a4f.js",
      "/static_shared/mathjax/newcommand.3ed90b8ad4b0c8062cd5.js",
      "/static_shared/mathjax/safe.48438015c1536976b6b7.js",
      "/static_shared/mathjax/lazy.c066af8a7aaa83a16cbc.js",
      "/static_shared/mathjax/empheq.c815940a3d4f84d783eb.js",
      "/static_shared/mathjax/color.e63944ab1f419699c81e.js",
      "/static_shared/mathjax/semantic-enrich.2b71fb880885444da334.js",
      "/static_shared/mathjax/cases.202aa18aaacb250011c5.js",
      "/static_shared/mathjax/amscd.dc8f77cedbf8de51491a.js",
      "/static_shared/mathjax/braket.65b4b7f29db35c005047.js",
      "/static_shared/mathjax/textcomp.00a5534767ecb8e8624a.js",
      "/static_shared/mathjax/assistive-mml.34fc258df88fce6606af.js",
      "/static_shared/mathjax/autoload.77c93dcd4844d4b6463c.js",
      "/static_shared/mathjax/require.998d51c8cb6d31f3f661.js",
      "/static_shared/mathjax/setoptions.c0d162588ee0fe50cb22.js",
      "/static_shared/mathjax/colortbl.988a12b2e93fd3b4b53c.js",
      "/static_shared/mathjax/latest.3a2c732224f3b1776a4c.js",
      "/static_shared/mathjax/configmacros.61e745ecfda313acc32b.js",
      "/static_shared/mathjax/boldsymbol.e2519c92a56eeca69603.js",
      "/static_shared/mathjax/centernot.de5508ef440dbc13ab2d.js",
      "/static_shared/mathjax/extpfeil.142c73571d44b4dac74f.js",
      "/static_shared/mathjax/unicode.a2e3818800f8987eea0b.js",
      "/static_shared/mathjax/cancel.7ad3f23c43c15c61f98a.js",
      "/static_shared/mathjax/bbox.3ecf8e797beac9e9ead8.js",
      "/static_shared/mathjax/html.0cfecb9e2f2d345f320f.js",
      "/static_shared/mathjax/tagformat.cafe7cac91878719326a.js",
      "/static_shared/mathjax/action.41adbdb60bb262202180.js",
      "/static_shared/mathjax/verb.796ce49ebaa3111ad849.js",
      "/static_shared/mathjax/upgreek.0279af5cbc9872a9f4c4.js",
      "/static_shared/mathjax/enclose.f2478c89a617ba353949.js",
      "/static_shared/mathjax/gensymb.c0f7e4d27fe1e6834c13.js",
      "/static_shared/mathjax/colorv2.10c4ac8f279b0e9358a8.js",
      "/static_shared/mathjax/noundefined.d4cb70e134ffd02ac0df.js",
      "/static_shared/mathjax/noerrors.84c73d74a69a837189b8.js"
    ],
    "json": [
      "/static_shared/mathjax/hi.aab74c0bc3ae0482b852.json",
      "/static_shared/mathjax/en.a9b138d89f24a4e78af5.json",
      "/static_shared/mathjax/de.6a6c1545e99fb7ff487d.json",
      "/static_shared/mathjax/it.be605cff0a369902a1d0.json",
      "/static_shared/mathjax/fr.4682df8d6acb3c7e8d07.json",
      "/static_shared/mathjax/sv.3eff416a426945fe1be5.json",
      "/static_shared/mathjax/nn.a85f21ce5637b76878f2.json",
      "/static_shared/mathjax/nb.d4a693a3d7c51cd59ce1.json",
      "/static_shared/mathjax/es.1e699b95de583f2e5b72.json",
      "/static_shared/mathjax/ca.cf3917c0adc60df0af0b.json",
      "/static_shared/mathjax/da.f6abbb386e35024a23f9.json",
      "/static_shared/mathjax/nemeth.24f16eb5ddc100d00ac5.json",
      "/static_shared/mathjax/mml3.sef.c3712f9719ff8fd2dfa5.json",
      "/static_shared/mathjax/base.2f9c71c6367b5a17eb45.json"
    ],
    "ttf": [
      "/static_shared/fonts/Cardo-Regular.ttf",
      "/static_shared/fonts/Cardo-Bold.ttf",
      "/static_shared/fonts/Cardo-Italic.ttf",
      "/static_shared/fonts/Helvetica-Light.ttf",
      "/static_shared/fonts/MaterialIcons-Regular.subset.ttf"
    ],
    "jpg": [
      "/static_shared/images/course_image.jpg",
      "/static_shared/images/ocw_logo.jpg"
    ],
    "png": [
      "/static_shared/images/ocw_logo_orange.png",
      "/static_shared/images/default_news.png",
      "/static_shared/images/video_thumbnail.png",
      "/static_shared/images/oeg_logo.png",
      "/static_shared/images/button_start.png",
      "/static_shared/images/audio_thumbnail.png",
      "/static_shared/images/mobile_video_thumbnail.png",
      "/static_shared/images/mit-ol.png",
      "/static_shared/images/video_play_overlay.png",
      "/static_shared/images/150d9d70c96b5f4386e5.png",
      "/static_shared/images/page_thumbnail.png",
      "/static_shared/images/Twitter.png",
      "/static_shared/images/LinkedIn.png",
      "/static_shared/images/Instagram.png",
      "/static_shared/images/Youtube.png",
      "/static_shared/images/Facebook.png"
    ],
    "svg": [
      "/static_shared/images/ocw_logo_white_v2.svg",
      "/static_shared/images/heart.svg",
      "/static_shared/images/expand.svg",
      "/static_shared/images/youtube.svg",
      "/static_shared/images/videojs_download.svg",
      "/static_shared/images/close_small.svg",
      "/static_shared/images/mit-logo-black-red.svg",
      "/static_shared/images/MIT-logo-red-gray-54x28.svg",
      "/static_shared/images/left_arrow.svg",
      "/static_shared/images/mit-logo-red.svg",
      "/static_shared/images/search.svg",
      "/static_shared/images/resource-download.svg"
    ],
    "woff": [
      "/static_shared/mathjax/MathJax_AMS-Regular.8c5473f40be1e271b35f.woff",
      "/static_shared/mathjax/MathJax_Main-Bold.ec22e0762cd55d9fcb83.woff",
      "/static_shared/mathjax/MathJax_Main-Regular.8c5439beba3f01241e88.woff",
      "/static_shared/mathjax/MathJax_Fraktur-Bold.374b2906529cb2a8d158.woff",
      "/static_shared/mathjax/MathJax_Fraktur-Regular.d379175560459eda3960.woff",
      "/static_shared/mathjax/MathJax_Main-Italic.7b3896ca3fbef758db3a.woff",
      "/static_shared/mathjax/MathJax_Math-BoldItalic.85b228bc0d50d5d8b04a.woff",
      "/static_shared/mathjax/MathJax_Math-Italic.85916f71942cc0883ab3.woff",
      "/static_shared/mathjax/MathJax_Math-Regular.0d54442637d18925f54c.woff",
      "/static_shared/mathjax/MathJax_Typewriter-Regular.5b9d001bddd151a72dcc.woff",
      "/static_shared/mathjax/MathJax_SansSerif-Bold.abaf4c9e83dbaf05fdc7.woff",
      "/static_shared/css/fonts/bbe2fbc76615c7cbd85ea132d0ff92e7.woff",
      "/static_shared/mathjax/MathJax_SansSerif-Italic.4238947244ed85c87b3f.woff",
      "/static_shared/mathjax/MathJax_SansSerif-Regular.0dd27dc8f931ca441afd.woff",
      "/static_shared/mathjax/MathJax_Script-Regular.9faee4c1f1002c39ae7e.woff",
      "/static_shared/mathjax/MathJax_Calligraphic-Bold.60ec5e1aad6265ebc57e.woff",
      "/static_shared/mathjax/MathJax_Calligraphic-Regular.95d40208de043ccfb7e8.woff",
      "/static_shared/mathjax/MathJax_Size1-Regular.7afc669b6d3e1e74c0b7.woff",
      "/static_shared/mathjax/MathJax_Size2-Regular.1715f238ef71d3ff3c90.woff",
      "/static_shared/mathjax/MathJax_Size4-Regular.ecf84a30e05c16f91074.woff",
      "/static_shared/mathjax/MathJax_Size3-Regular.84c7037e9001a88f1c3d.woff",
      "/static_shared/fonts/MaterialIcons-Regular.subset.woff",
      "/static_shared/mathjax/MathJax_Zero.fbe2a3a1cc54d8c591fa.woff",
      "/static_shared/mathjax/MathJax_Vector-Regular.cc199a82f902b005545c.woff",
      "/static_shared/mathjax/MathJax_Vector-Bold.b1da3c213f29eb271a3a.woff"
    ],
    "css": "/static_shared/css/948.150d9.css",
    "woff2": [
      "/static_shared/css/fonts/22d194ee0b49f6e9d633ddc302f108b1.woff2",
      "/static_shared/fonts/MaterialIcons-Regular.subset.woff2",
      "/static_shared/fonts/MaterialIconsRound-Regular.subset.woff2"
    ],
    "gif": [
      "/static_shared/images/a_logo_17.gif",
      "/static_shared/images/mp_logo.gif",
      "/static_shared/images/inacessible.gif"
    ]
  }
}
```

So, Webpack is generating assets without a hash and those filenames are getting written into `webpack.json`, then the hash is being added by Hugo at build time to the URLs referencing these files, thus referencing a file that doesn't exist.

Reply:
If you want to run this yourself, set `OCW_HUGO_THEMES_BRANCH=ibrahim/1212/hash-static-assets` in your `ocw-studio` env, restart your containers and then run:

```
docker-compose exec web ./manage.py upsert_theme_assets_pipeline
docker-compose exec web ./manage.py backpopulate_pipelines
```

Then when you publish any site it will be using your branch.

Reply:
Doing some further research into this, I looked at the deployed folder in Minio and found that the Hugo build had deposited a `static_shared` folder inside of the course folder which contained the hashed fonts and images:

![image](https://github.com/mitodl/ocw-hugo-themes/assets/12089658/a458a9bf-3f1c-40d7-8b41-13a0921220ef)
![image](https://github.com/mitodl/ocw-hugo-themes/assets/12089658/fbaea0a9-0121-4836-9cdb-36e67e4d0be7)

So, it seems Hugo fingerprinting won't work for assets in `static_shared`, as they are meant to be shared across all courses at the root of the site. They will either need to be hashed by Webpack instead, or a "dummy" Hugo build will need to be run as part of the theme assets pipeline that generates the fingerprinted assets and puts them in `static_shared` at the root of the site.

I put up a WIP PR here that demonstrates what we would need to do to generate the fingerprinted assets during the theme assets build here: https://github.com/mitodl/ocw-studio/pull/2006. I was able to verify that running the pipeline worked for me. Once I ran the theme assets pipeline with these changes, the fingerprinted assets were generated and copied over into the root of the site.